### PR TITLE
Draft version of Ethereum development tips post

### PR DESCRIPTION
**Pull request by misterzero**

### CODE REVIEW

File: posts/ethereum-development-tips.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+(source: https://coin24.fr/2018/03/07/faut-vraiment-acheter-ethereum-2018/)
+
+# Summary
+
+When we arrived at Ippon USA for our summer internship, we had little knowledge about blockchains. The goal of this article is to give you the basic tools one can use to build a decentralized application, and some tips about the language we used: Solidity. Of course our list is not exhaustive, and is just here to offer you a peek at Ethereum development.
+
+# Introduction to Blockchains and Ethereum
+
+Let’s start with the Wikipedia definition:
+"A blockchain, originally <b>block chain</b>, is a growing list of records, called blocks, which are linked using cryptography. Blockchains which are readable by the public are widely used by cryptocurrencies. Private blockchains have been proposed for business use."
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: use `>` to quote text, e.g.:

> quoted text

File: posts/ethereum-development-tips.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+Perfect, you know the different Ethereum blockchains and how to connect yourself and your applications to them. Now you just need to actually code and deploy them. You can code it like other project on an environment such as Visual Studio Code which has .sol extensions (for Solidity, or directly use Remix (remix.ethereum.org). It has several major uses. Its compiler will directly say if you smart contract can work and all the possible warnings attached to it. After, you can deploy your contract on the Ethereum net you want using one of the web3.js solutions aforementioned. But it doesn’t stop there! You can directly interact with the deployed contracts with simple buttons to call the different functions, and you will receive all the informations needed about the results or the transactions in the Remix console.
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: "You can code it like other project" -> not sure of the meaning

Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: use Remix (remix.ethereum.org) -> put a proper link

Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: .sol -> put it between backquotes to make it look like code

File: posts/ethereum-development-tips.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+{{< responsive-figure src="/images/2018/08/blockchain-interns-08.png" >}}
+4. Metamask: Metamask is a Chrome/Firefox plugin using web3.js. It gives a convenient way to connect to the different Ethereum Blockchains (mainnet and testnets). It stores your public and private key and allow you to call and send transaction to smart contracts without running a full ethereum node. It is the most user friendly way to connect to Ethereum.
+
+{{< responsive-figure src="/images/2018/08/blockchain-interns-09.png" >}}
+5. geth and Mist: geth is a tool that enables you to run a local or a public blockchain node. It can be used with Mist, a browser that makes it easy to interact with smart contracts and your accounts.
+
+{{< responsive-figure src="/images/2018/08/blockchain-interns-10.png" >}}
+6. Remix: Remix is an online Solidity IDE allowing you to compile your code, and deploy it on the mainnet or testnets with the previous environments : Metamask, Ganache, geth… Afterwards you can directly interact with it with simple buttons, and receive all the informations needed about the results and the transactions in the Remix console.
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: remove space before `:`

File: posts/ethereum-development-tips.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+{{< responsive-figure src="/images/2018/08/blockchain-interns-11.png" >}}
+7. Dai: Dai is a stablecoin i.e. an ERC20 token which price is stable and equal to the value of the USD. It is the only stablecoin where users do not need to trust a centralized instance. Implementing it on a decentralized application means a user balance value will not fluctuate.
+
+{{< responsive-figure src="/images/2018/08/blockchain-interns-12.png" >}}
+8. Etherscan: Etherscan is a “Blockchain explorer”, providing you with all the informations you want about a wallet, contract, or a transaction on the blockchain! You just need to give it an address to look at.
+
+{{< responsive-figure src="/images/2018/08/blockchain-interns-13.png" >}}
+9. ZeppelinOS: We were on the brink of using ZeppelinOS, a development platform with a clear promise : easily upgradable smart contracts with the help of proxy contracts.
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: remove space before `:`

File: posts/ethereum-development-tips.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+# Solidity tips
+
+{{< responsive-figure src="/images/2018/08/blockchain-interns-15.png" >}}
+
+If you made it this far, you are really interested on how the Ethereum blockchain works, or you want to get to it and develop in Solidity! There is a lot to know about Solidity: although it was made with a really high-level appearance close to Java, the aforementioned specifics of blockchain and smart contracts differentiate it a lot. A lot of trial and error was necessary for us and do not hesitate to ask for help on StackExchange or all your favorite communities, Ethereum development is very active.
+
+The basic idea is that a contract is a complete package, containing state variables, functions, functions modifiers, events, struct types, and enum types. You can also inherit from other contracts. It looks really similar to a Java class, it even also has a constructor if you want to initialize some variables. You can also easily access to the transaction attributes :
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: remove space before `:`

Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: functions modifiers -> function modifiers

File: posts/ethereum-development-tips.md [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+-msg.data (bytes): complete calldata
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: add a space after `-`

File: posts/ethereum-development-tips.md [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: remove extra lines between list items

File: posts/ethereum-development-tips.md [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+# Summary
+
+When we arrived at Ippon USA for our summer internship, we had little knowledge about blockchains. The goal of this article is to give you the basic tools one can use to build a decentralized application, and some tips about the language we used: Solidity. Of course our list is not exhaustive, and is just here to offer you a peek at Ethereum development.
+
+# Introduction to Blockchains and Ethereum
+
+Let’s start with the Wikipedia definition:
+"A blockchain, originally <b>block chain</b>, is a growing list of records, called blocks, which are linked using cryptography. Blockchains which are readable by the public are widely used by cryptocurrencies. Private blockchains have been proposed for business use."
+
+To put it in layman’s terms, a blockchain is simply a transaction record that is decentralized. It means anybody can store and run a copy of the blockchain (becoming a “node”) and therefore read it.
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: replace all special quotes `“` with regular quotes `"`

File: posts/ethereum-development-tips.md [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    contract SimpleMarket {  
+        event HighestSell(address seller, uint amount); // Event  
+
+        function sell() public {  
+        // ...  
+        emit HighestSell(msg.sender, msg.value); // Triggering event  
+        }
+    }
+
+The types used in Solidity are pretty much the ones you know and are fairly close to C. String are dynamic arrays of characters, and as such can be tricky to manage. But you may be surprised to not see float, as they do not exist. You will need to be cautious and use multipliers! The types developers seem to affectionate the most are uint256 as they offer the largest numbers (2**256 integers), and addresses, as they are used for most objects such as accounts and contracts.
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: 2**256 -> 2^256

File: posts/ethereum-development-tips.md [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+You have to know if the object you are defining in your function needs to be “memory” (stored temporarily) or “storage” (knowing that it will be stored in the blockchain). For example, a dynamic array has to be storage, so you may need to find a way to know the exact number of object it should need.
+
+Other oddities may include structures not being able to be defined recursively, or more importantly, all state variables being initialized to their default value. This means that a boolean will be automatically false, and also that a mapping (Solidity’s widely used equivalent of a hashmap) will be initialized for every key at the byte 0. This means that you cannot iterate on the keys!
+
+    LinkedList.Data  storage list;
+
+    uint256[]  memory _parts;
+
+Complex storage structures can be useful, such as the Linked List we needed to implement. By using a mapping from a uint id to a Node structure (with the data we want, the information to it will be sorted by, and the id of the next Node) and keeping the id of the head of the list, we could make an easy sortable list where we can add and delete nodes.
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: the information to it will be sorted by -> not sure of the meaning

Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: easy sortable list -> easily sortable list (?)

File: posts/ethereum-development-tips.md [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+
+One of the burden of Solidity is how to code a secure smart contract. Writing a contract that behaves as you expect is quit easy, however making sure it won’t later act strangely is hader. Solidity, is a constantly evolving language, but can still fall short of being a secure language for creating decentralized application.
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: quit -> quite

Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: hader -> harder

File: posts/ethereum-development-tips.md [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+
+One of the tools provided by Solidity is for Error handling. You will probably want to have quite a few “Require” functions used throughout your code, as they will “Revert” (which means cancelling the execution of the transaction and revert back any gas used to the sender) when you need them to. They can be used to check users balance before transfers for instance.
+
+Now let’s go over a few problems or considerations you may encounter while coding in Solidity (well, that we have encountered or seen in solidity hacking challenges). A lot of them are really specific to the way blockchain works.
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: solidity -> Solidity

File: posts/ethereum-development-tips.md [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+
+The most common and biggest problem to check out is the classic reentrancy problem. For example, you do not want a user to be able to call two times a function (or two different functions) that will give him tokens before depleting his balance!
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: Not sure this is reentrancy. Reentrancy is the capacity of a function to call itself (re-enter).

File: posts/ethereum-development-tips.md [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+
+            if (msg.sender.send(shares[msg.sender]))
+
+            shares[msg.sender] =  0;
+        }
+    }
+
+Close to this problem and more specific to blockchain, is the fact that every transaction being public, a malicious user could see a transaction he’s interested in, and before it is mined, do the exact same transaction but with way more gas, making his transaction be processed faster!
+
+If the code of to a soon to be deployed app is public, why not do more, and precompute its future address to send ETH to it before, allowing unexpected events? You can precompute a lot actually, and doing random functions in your code that use block number or timestamp (the time a block is mined) is not safe as it is pretty easy to figure out!
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: of to a soon -> of a soon (?)

Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: soon to be deployed -> soon-to-be-deployed

File: posts/ethereum-development-tips.md [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+-msg.sender (address): sender of the message (current call)
+
+-msg.sig (bytes4): first four bytes of the calldata (i.e. function identifier)
+
+-msg.value (uint): number of wei sent with the message (1ETH = 10**18 wei)
+
+A function, if it changes any state variable, will cost gas fees to the user or contract calling it.
+
+
+    pragma solidity ^0.4.0;
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: use triple backquotes for all code blocks, and use `solidity` as the language

File: posts/ethereum-development-tips.md [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+
+There is a special function that exists in a contract to address the need for floating point. It does not have a name, and will be executed when no other appropriate function is found. It is called the fallback function, and can be used for quite a few interesting hacking tricks. However as we found out, you can not close it up completely, “just to be sure”, as you need to have a payable fallback function if you want it to receive ETH from another contract!
+
+    function() public payable { //Fairly open now!
+
+    }
+
+Security in Solidity, as in every language, is a difficult system of checks and balances between features and protection.
+
+Of course we had to link the official solidity docs, they are well documented and updated regularly:
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: put these 3 links on some text or on the title of the targeted page

File: posts/ethereum-development-tips.md [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+
+[https://solidity.readthedocs.io/en/v0.4.24/](https://solidity.readthedocs.io/en/v0.4.24/)
+
+A classic game of increasingly difficult solidity hacking challenges. It gets extremely tough!
+
+[https://ethernaut.zeppelin.solutions/](https://ethernaut.zeppelin.solutions/)
+
+A coding contest that highlights a lot of mistakes or malicious attacks possible in Solidity. There’s lot of exemples to check out on the github!
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: github -> GitHub

Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: on the github -> either:
- on the GitHub account
- on GitHub

File: posts/ethereum-development-tips.md [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+[https://medium.com/@weka/announcing-the-winners-of-the-first-underhanded-solidity-coding-contest-282563a87079](https://medium.com/@weka/announcing-the-winners-of-the-first-underhanded-solidity-coding-contest-282563a87079)
+
+A collection of "best practices" in the Ethereum development world by Consensys, a major actor in the Ethereum market.
+
+https://consensys.github.io/smart-contract-best-practices/  
+
+# Conclusion
+
+We hope this small introduction to Ethereum development has helped you. We also would like to sincerely thank the entire team of Ippon USA who hosted two French interns in Washington DC, and especially our tutor Chris Lumpkin, Dagmawi Mengistu, Priya Rajanna and Matt Reed our co-workers who welcomed and showed us a priceless glimpse of the American melting pot culture.
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: Washington DC -> Washington, DC

Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: our tutor Chris Lumpkin, Dagmawi Mengistu, Priya Rajanna and Matt Reed our co-workers -> confusing

File: posts/ethereum-development-tips.md [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+- ZeppelinOS
+- web3
+- Metamask
+- geth
+date: 
+title: "Ethereum Development Tips"
+image: 
+---
+
+{{< responsive-figure src="/images/2018/08/blockchain-interns-01.png" >}}
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: this image be referenced in metadata

File: posts/ethereum-development-tips.md [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+(source: https://coin24.fr/2018/03/07/faut-vraiment-acheter-ethereum-2018/)
+
+# Summary
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: there will already be a title level 1, so this can be removed to make `When we arrived...` the first text block

File: posts/ethereum-development-tips.md [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+
+To put it in layman’s terms, a blockchain is simply a transaction record that is decentralized. It means anybody can store and run a copy of the blockchain (becoming a “node”) and therefore read it.
+
+There is not a single blockchain, but different ones for different purposes. The Bitcoin blockchain is the most famous one of all, but there is also the widely popular Ethereum blockchain which we use, and many others. These are public, but you can as easily create a private blockchain, which won’t be available to everyone.
+
+As on a classic ledger, you can only add data to a blockchain. To add new pages to the ledger or new blocks to a blockchain “miners” have to solve complicated and costly algorithm problems. Your transaction will be validated after a few blocks have been mined. This is why transactions cost a fee to reward miners.
+
+Finally, security is a major asset of blockchains, with the help of high-end cryptography and the distributed network effect.
+
+{{< responsive-figure src="/images/2018/08/blockchain-interns-02.png" >}}
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: Use `text`

Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: /images/2018/08/ -> /images/2018/09/

Comment (misterzero) [reply posts/ethereum-development-tips.md]: In that case, the README needs to be updated... I will submit PR for README when I have time.

File: posts/ethereum-development-tips.md [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
+- ZeppelinOS
+- web3
+- Metamask
+- geth
+date: 
+title: "Ethereum Development Tips"
+image: "/images/2018/09/blockchain-interns-01.png"
+---
+
+![Ethereum Development Tips](/images/2018/09/blockchain-interns-01.png)
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: we should remove this from here now, otherwise it will be displayed twice: once as the main image, once with this inline image

File: posts/ethereum-development-tips.md [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
+To put it in layman’s terms, a blockchain is simply a transaction record that is decentralized. It means anybody can store and run a copy of the blockchain (becoming a "node”) and therefore read it.
+
+There is not a single blockchain, but different ones for different purposes. The Bitcoin blockchain is the most famous one of all, but there is also the widely popular Ethereum blockchain which we use, and many others. These are public, but you can as easily create a private blockchain, which won’t be available to everyone.
+
+As on a classic ledger, you can only add data to a blockchain. To add new pages to the ledger or new blocks to a blockchain "miners” have to solve complicated and costly algorithm problems. Your transaction will be validated after a few blocks have been mined. This is why transactions cost a fee to reward miners.
+
+Finally, security is a major asset of blockchains, with the help of high-end cryptography and the distributed network effect.
+
+![Ethereum](/images/2018/09/blockchain-interns-02.png)
+(source: https://medium.com/the-mission/ethereum-will-be-the-next-facebook-a-fundamental-analysis-40bd95fde2a4)
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: can we put a link either like `[source](url)` or `source: [title of the targeted page](url)` so that we don't display a URL?

File: posts/ethereum-development-tips.md [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+- msg.data (bytes): complete calldata
+- msg.gas (uint): remaining gas - deprecated in version 0.4.21 and to be replaced by gasleft()
+- msg.sender (address): sender of the message (current call)
+- msg.sig (bytes4): first four bytes of the calldata (i.e. function identifier)
+- msg.value (uint): number of wei sent with the message (1ETH = 10^18 wei)
+
+A function, if it changes any state variable, will cost gas fees to the user or contract calling it.
+
+```solidity
+    pragma solidity ^0.4.0;
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: indentation should be removed now

File: posts/ethereum-development-tips.md [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
+```solidity
+    function() public payable { //Fairly open now!
+
+    }
+```
+
+Security in Solidity, as in every language, is a difficult system of checks and balances between features and protection.
+
+Of course we had to link the official solidity docs, they are well documented and updated regularly:
+- [https://solidity.readthedocs.io/en/v0.4.24/](https://solidity.readthedocs.io/en/v0.4.24/)
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: same as above: text of the link should be the title of the targeted page

File: posts/ethereum-development-tips.md [hunk 26]
```diff
@@ -0,0 +1,260 @@
...
+
+A coding contest that highlights a lot of mistakes or malicious attacks possible in Solidity. There’s lot of exemples to check out on GitHub!
+- [https://medium.com/@weka/announcing-the-winners-of-the-first-underhanded-solidity-coding-contest-282563a87079](https://medium.com/@weka/announcing-the-winners-of-the-first-underhanded-solidity-coding-contest-282563a87079)
+
+A collection of "best practices" in the Ethereum development world by Consensys, a major actor in the Ethereum market.
+- [https://consensys.github.io/smart-contract-best-practices/](https://consensys.github.io/smart-contract-best-practices/)
+
+# Conclusion
+
+We hope this small introduction to Ethereum development has helped you. We also would like to sincerely thank the entire team of Ippon USA who hosted two French interns in Washington DC: our mentor Chris Lumpkin, our co-workers Dagmawi Mengistu, Priya Rajanna and Matt Reed who all welcomed us and showed us a priceless glimpse of the American melting pot culture.
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: small -> do we want to keep this term, or maybe replace with "short"?

File: posts/ethereum-development-tips.md (hunk 23 above)
Comment (misterzero) [reply posts/ethereum-development-tips.md]: I'm removing the attribution, this is a stock Ethereum image and the article doesn't claim any copyright.

File: posts/ethereum-development-tips.md [hunk 27]
```diff
@@ -0,0 +1,255 @@
...
+
+```solidity
+function() public payable { //Fairly open now!
+
+}
+```
+
+Security in Solidity, as in every language, is a difficult system of checks and balances between features and protection.
+
+- [https://solidity.readthedocs.io/en/v0.4.24/](Official solidity docs)
```
Comment (aseigneurin) [inline posts/ethereum-development-tips.md]: the other way around: `[text](link)`, not `[link](text)` :)
